fluture.js

```diff
@@ -376,544 +404,927 @@
+      }
+      while(!state.done){
+        isSync = null;
+        const m = _this._iterate(Next, Done, state.value);
+        check$chainRec$f(m, _this._iterate, i, state.value);
+        cancel = m._f(rej, Future$chainRec$res);
+        if(isSync === true){
+          continue;
+        }else{
+          isSync = false;
+          return;
+        }
+      }
+      res(state.value);
+    }(Next(_this._init)));
+    return function Future$chainRec$cancel(){ cancel() };
   }
 
-  function Future$map(f){
-    check$map(this, f);
-    const _this = this;
-    return new FutureClass(false, function Future$map$fork(rej, res){
-      return _this._f(rej, function Future$map$res(x){
-        res(f(x));
-      });
-    });
+  ChainRec.prototype.toString = function ChainRec$toString(){
+    return `Future.chainRec(${showf(this._iterate)}, ${show(this._init)})`;
   }
 
-  function Future$mapRej(f){
-    check$mapRej(this, f);
-    const _this = this;
-    return new FutureClass(false, function Future$mapRej$fork(rej, res){
-      return _this._f(function Future$mapRej$rej(x){
-        rej(f(x));
-      }, res);
-    });
+  //----------
+
+  function CachedFuture(pure){
+    this._pure = pure;
+    this._cancel = noop;
+    this._queue = [];
+    this._queued = 0;
+    this._value = null;
+    this._state = 0;
   }
 
-  function Future$bimap(f, g){
-    check$bimap(this, f, g);
+  CachedFuture.STATE = {
+    0: 'cold',
+    1: 'pending',
+    2: 'rejected',
+    3: 'resolved'
+  };
+
+  CachedFuture.prototype = Object.create(Future.prototype);
+
+  CachedFuture.prototype._addToQueue = function CachedFuture$addToQueue(rej, res){
     const _this = this;
-    return new FutureClass(false, function Future$bimap$fork(rej, res){
-      return _this._f(function Future$bimap$rej(x){
-        rej(f(x));
-      }, function Future$bimap$res(x){
-        res(g(x));
-      });
-    });
+    const i = _this._queue.push({2: rej, 3: res}) - 1;
+    _this._queued = _this._queued + 1;
+    return function CachedFuture$removeFromQueue(){
+      if(_this._state > 1) return;
+      _this._queue[i] = undefined;
+      _this._queued = _this._queued - 1;
+      if(_this._queued === 0) _this.reset();
+    };
   }
 
-  function Future$ap(m){
-    check$ap(this, m);
-    const _this = this;
-    return new FutureClass(false, function Future$ap$fork(g, res){
-      let _f, _x, ok1, ok2, ko;
-      const rej = x => ko || (ko = 1, g(x));
-      const c1 = _this._f(rej, function Future$ap$resThis(x){
-        if(!ok1) return void (ok2 = 1, _x = x)
-        check$ap$f(_f);
-        res(_f(x));
-      });
-      const c2 = m._f(rej, function Future$ap$resThat(f){
-        if(!ok2) return void (ok1 = 1, _f = f);
-        check$ap$f(f);
-        res(f(_x));
-      });
-      return function Future$ap$cancel(){ c1(); c2() };
-    });
+  CachedFuture.prototype._drainQueue = function CachedFuture$drainQueue(){
+    const q = this._queue, l = q.length, s = this._state, v = this._value;
```

Comment:
No need  to name these such small names. I actually just did this to move property-access to outside of the loop.